File: backend/app/relations.py
"""관계 predicate → 6그룹 매핑 (프론트 src/lib/relations.ts 의 백엔드 미러, SSOT)."""
from __future__ import annotations

# Neo4j 관계타입 → 화면 6그룹
PREDICATE_TO_GROUP: dict[str, str] = {
    "SUPPLIES": "supply", "CUSTOMER_OF": "supply",
    "COMPETES_WITH": "compete",
    "PARTNERS_WITH": "partner", "JV_WITH": "partner", "LICENSES": "partner",
    "INVESTS_IN": "invest", "ACQUIRES": "invest",
    "IS_SUBSIDIARY_OF": "govern", "IS_MAJOR_SHAREHOLDER_OF": "govern", "AFFILIATED_WITH": "govern",
    "LITIGATION": "dispute",
}

# 회사↔회사 그래프에 그릴 관계타입 (인물 EXECUTIVE_OF, 제품 DEVELOPS 는 그래프 제외 → 패널)
COMPANY_REL_TYPES: list[str] = list(PREDICATE_TO_GROUP)

# 방향성 있는 관계 (화살표 표시)
DIRECTED: set[str] = {
    "SUPPLIES", "CUSTOMER_OF", "INVESTS_IN", "ACQUIRES",
    "IS_SUBSIDIARY_OF", "IS_MAJOR_SHAREHOLDER_OF",
}

# 그룹 → predicate 목록 (근거 조회 시 그룹으로 역매핑)
GROUP_TO_PREDICATES: dict[str, list[str]] = {}
for _p, _g in PREDICATE_TO_GROUP.items():
    GROUP_TO_PREDICATES.setdefault(_g, []).append(_p)

# 1차 시드 3사 (노드 강조용)
SEED_CORPS: dict[str, str] = {
    "00126380": "삼성전자",
    "00164779": "SK하이닉스",
    "00161383": "한미반도체",
}

File: frontend/src/lib/relations.ts
import type { RelationGroup } from '../types'

// 관계 6그룹의 단일 출처(SSOT): 라벨·색(엣지 전용 의미색)·백엔드 predicate 매핑.
export interface RelationMeta {
  key: RelationGroup
  label: string
  color: string
  predicates: string[]
}

// 뮤트 팔레트 — 동일 채도·명도로 묶어 형형색색 X, 차분·고급 (Carbon/Cloudscape 데이터비주얼 가이드)
export const RELATION_GROUPS: RelationMeta[] = [
  { key: 'supply', label: '공급망', color: '#5E9BD1', predicates: ['SUPPLIES', 'CUSTOMER_OF'] },
  { key: 'compete', label: '경쟁', color: '#D9737A', predicates: ['COMPETES_WITH'] },
  { key: 'partner', label: '협력', color: '#5FB39C', predicates: ['PARTNERS_WITH', 'JV_WITH', 'LICENSES'] },
  { key: 'invest', label: '투자·인수', color: '#9B86CB', predicates: ['INVESTS_IN', 'ACQUIRES'] },
  { key: 'govern', label: '지배구조', color: '#8E99A8', predicates: ['IS_SUBSIDIARY_OF', 'IS_MAJOR_SHAREHOLDER_OF', 'AFFILIATED_WITH'] },
  { key: 'dispute', label: '분쟁', color: '#E0A45E', predicates: ['LITIGATION'] },
]

export const GROUP_COLOR = Object.fromEntries(
  RELATION_GROUPS.map((g) => [g.key, g.color]),
) as Record<RelationGroup, string>

export const GROUP_LABEL = Object.fromEntries(
  RELATION_GROUPS.map((g) => [g.key, g.label]),
) as Record<RelationGroup, string>
